documentation/secure_storage_rpmb.md

```diff
@@ -0,0 +1,108 @@
+# RPMB Secure Storage
+
+## Introduction
+
+There are currently two secure storage implementations in OP-TEE:
+
+- The first one relies on the normal world (REE) file system. It is described in
+[secure_storage.md](secure_storage.md) and is the default implementation.
+- The second one makes use of the Replay Protected Memory Block (RPMB) partition
+of an eMMC device, and is enabled by setting `CFG_RPMB_FS=y`.
+
+It is not currently possible to use the normal world filesystem and the RPMB
+implementations simultaneously. This document describes the RPMB implementation.
+The architecture is depicted below.
+
+```
+            NORMAL WORLD           :            SECURE WORLD
+                                   :
+ U        tee-supplicant           :        Trusted application
+ S           (rpmb.c)              :        (secure storage API)
+ E         ^          ^            :                  ^
+ R         |          |            :~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~
+ ~~~~~~~ ioctl ~~~~~~~|~~~~~~~~~~~~:                  v
+ K         |          |            :               OP-TEE
+ E         v          v            :         (tee_svc_storage.c)
+ R    eMMC driver   OP-TEE driver  : (tee_rpmb_fs.c, tee_fs_key_manager.c)
+ N                           ^     :            (tee_rpmb.c)
+ E                           |     :                  ^
+ L                           |     :                  |
+                             |     :                  |
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~~~~~|~~~~~~~~~~~~~~~~~~~~
+                             v                        v
+                           Secure monitor / EL3 firmware
+```
+
+## The Secure Storage API
+
+This part is common with the REE-based filesystem. The interface between the
+system calls in [core/tee/tee_svc_storage.c](../core/tee/tee_svc_storage.c) and
+the RPMB filesystem is the **tee_file_operations** structure `tee_file_ops`.
+
+## The RPMB filesystem
+
+The FS implementation is mainly in [core/tee/tee_rpmb_fs.c](../core/tee/tee_rpmb_fs.c).
+
+The RPMB partition is divided in three parts:
+
+- The first 128 bytes are reserved for partition data (struct
+**rpmb_fs_partition**).
+- At offset 512 is the File Allocation Table (FAT). It is an array of
+struct **rpmb_fat_entry** elements, one per file. The FAT grows dynamically as
+files are added to the filesystem. Among other things, each entry has the start
+address for the file data, its size, and the filename.
+- Starting from the end of the RPMB partition and extending downwards is the
+file data area.
+
+Space in the partition is allocated by the general-purpose allocator functions:
+`tee_mm_alloc()` and `tee_mm_alloc2()`.
+
+## Device access
+
+The device access layer is in [core/arch/arm/tee/tee_rpmb.c](../core/arch/arm/tee/tee_rpmb.c).
+
+There is no eMMC controller driver in OP-TEE. The device operations, such as:
```

Comment:
I will update this section. The RPMB security key is indeed *not* the SSK mentioned below, so it has to be better documented.
As for the write counter: it is read on startup and kept in the RPMB context (struct tee_rpmb_ctx), i.e., `rpmb_ctx->wr_cnt`. I will clarify this also.